peps/pep-0797.rst

```diff
@@ -0,0 +1,672 @@
+------------------------------------
+
+Once again, this should already be true for all extensions.
+Since Python 3.4 (:pep:`442`), :c:member:`~PyTypeObject.tp_finalize` alongside
+:c:func:`PyObject_CallFinalizerFromDealloc` has been the correct way to run
+cleanup code for an extension type. Immortal object deallocation requires that
+:c:member:`~PyTypeObject.tp_dealloc` will not try and run Python code, and instead
+defer it to :c:member:`~PyTypeObject.tp_finalize`, for both GC and non-GC types.
+
+As long as :c:func:`PyObject_CallFinalizerFromDealloc` is used in the destructor,
+it is completely safe to run cleanup code, because it will just end up no-oping
+due to the de-duplication from :c:func:`PyObject_CallFinalizer` (the actual
+finalizer call will have been called earlier during immortal object cleanup;
+it's just that destruction is one of the last steps in immortal object
+deallocation).
+
+Using the Object or Memory Domain is a Must
+-------------------------------------------
+
+For an object immortal on the heap, the object domain (:c:func:`PyObject_Malloc`
+or :c:macro:`PyObject_New`) or the memory domain (:c:func:`PyMem_Malloc`)
+must be used. Otherwise, the interpreter will not be able to hijack the
+object's allocator and will result in use-after-free violations during
+deallocation. Note that in 3.13, using the object domain is already a
+:ref:`hard requirement <python:allocator-domains>` for the free threading
+garbage collector.
+
+    The free-threaded build requires that only Python objects are allocated
+    using the “object” domain and that all Python objects are allocated using
+    that domain. This differs from the prior Python versions, where this was
+    only a best practice and not a hard requirement.
+
+Reference Cycles Must be Clearable and Traversable
+--------------------------------------------------
+
+In order for immortal object deallocation to be safe, an object that can have
+reference cycles must be able to clear those reference cycles via
+their :c:member:`~PyTypeObject.tp_clear` slot.
+
+Currently, most extension types that hold a :term:`strong reference` to another
+Python object will be able to properly clear references through the garbage
+collector. In rare cases, it's possible that a buggy extension type could miss
+traversing a strong reference. For immortal object deallocation to work
+correctly, Python must be able to identify reference cycles on immortal objects.
+Prior to this proposal, reference cycles that aren't traversed by the garbage
+collector will just leak, but now if they're immortalized, they
+will crash. There's not all that much that can be done about this problem; it's
+a bug anyway to have these hidden reference cycles, but it just becomes more of
+an issue with this PEP.
+
+Immortal Object Deallocation
+****************************
+
+The hardest part of immortalizing arbitrary objects is dealing with
+deallocating them. The "intuitive" way of doing it would be to go
```

Comment:
This repeats the previous section, "*The hardest part about immortality is the deallocation*". I'd suggest that prose & the problem statement should be moved to *Motivation* & *Rationale*.